### PR TITLE
Push the logic of VersionedDao into type classes

### PR DESCRIPTION
**Pull request by alexwlchan**
The release note has the hierarchy; general gist is that this opens the door to easier testing if you use a VersionedDao implementation.

### CODE REVIEW

File: RELEASE.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+trait Dao[Ident, T] {
+  def get(id: Ident): Try[Option[T]]
+  def put(t: T): Try[T]
+}
+
+// Put conditions on the put() operation
+trait ConditionalUpdateDao[Ident, T] extends Dao[Ident, T]
+
+// The traditional VersionedDao
+trait VersionedDao[T] {
```
Comment (kenoir) [inline RELEASE.md]: This is a `ConditionalDao` right? It knows nothing about versioning.

File: storage/src/main/scala/uk/ac/wellcome/storage/ConditionalUpdateDao.scala [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package uk.ac.wellcome.storage
+
+trait ConditionalUpdateDao[Ident, T] extends Dao[Ident, T]
```
Comment (kenoir) [inline storage/src/main/scala/uk/ac/wellcome/storage/ConditionalUpdateDao.scala]: Interesting. Maybe we should talk through how the underlying type system here should work. It's practically probably just a naming rehash, and perhaps consolidating existing type classes.

File: storage/src/main/scala/uk/ac/wellcome/storage/Dao.scala [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package uk.ac.wellcome.storage
+
+import scala.util.Try
+
+trait Dao[Ident, T] {
```
Comment (kenoir) [inline storage/src/main/scala/uk/ac/wellcome/storage/Dao.scala]: 👏 That's what it is!

File: storage/src/main/scala/uk/ac/wellcome/storage/SerialisationStrategy.scala [hunk 4]
```diff
@@ -1,4 +1,4 @@
-package uk.ac.wellcome.storage.type_classes
```
Comment (kenoir) [inline storage/src/main/scala/uk/ac/wellcome/storage/SerialisationStrategy.scala]: Perfect. This namespace needed throwing away.

File: RELEASE.md (hunk 1 above)
Comment (alexwlchan) [reply RELEASE.md]: No, and I’ve leaked a bit too much into the release note here – the VersionedDao bumps the version and then calls the ConditionalUpdateDao to store the record.

Possibly there’s a fuzziness here; let’s talk about it later.

**Pull request merged**